### ISSUE TITLE
Allow season packs to be grabbed when all episodes have not aired.

### ISSUE THREAD

**Issue by obsidianwings** [Sonarr/Sonarr#6987]
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Is your feature request related to a problem? Please describe

Currently sonarr will not grab a season pack when all episodes have not aired, I think this may have made sense a few years ago when it was implemented, but now season packs are available more often straight away, maybe not so much anymore. This is especially prevalent with british tv shows where they air on certain dates on the tv and therefore tvdb, but they release the whole season on their web player the first day. Over the past few days this has been an issue for Mr Bigstuff and The Jetty, though I've had many others in the past.

### Describe the solution you'd like

I would like sonarr to either be changed so it will grab season packs as soon as they are available, or if fake packs on untrustworthy trackers is an issue simply give us a settings tickbox, perhaps with a warning, for those of us who trust our trackers not to have fake goods to allow season packs to be grabbed before all episodes have aired. Thanks for your time and consideration.

### Describe alternatives you've considered

The only alternative I can think of is what I am currently doing, trying to memorise which shows will be likely to drop the whole season early and perform a manual search and grab on them.

**Comment by bakerboy448** [comment]
perhaps allowing season packs if they dropped within X days of the first episode would be a solid middle ground?

Agree that removing the check is going to cause more problems than not, so a more nuanced feature would be warranted

**Comment by obsidianwings** [comment]
I'm open to the idea of putting in a longer time frame like that. The problem I see with that though is to cover the kind of shows I'm talking about all the time it would really have to be a matter of weeks. I'm thinking if removing the check would cause issues perhaps that period being weeks would also cause issues?

If so then I'm leaning towards it being optional with the default being as it is now being the best solution. Maybe hidden in advanced settings. Could cause more support issues with people enabling it that maybe shouldn't though I suppose.

**Comment by markus101** [comment]
This is really the inverse of https://github.com/Sonarr/Sonarr/issues/5312 where we don't have any restrictions and the recommendation is to just get better indexers.

Similar to https://github.com/Sonarr/Sonarr/issues/6378 this shouldn't necessarily be a global setting so having it part of Release Profiles could make sense, but the problem there is you can have 0 or more Release Profiles applied, to a series, unlike Delay Profiles where only one is applied and tags can be used to control the order.

It's a reasonable request, but is going to require some ideas on how to solve it.

**Comment by obsidianwings** [comment]
Thanks for considering it if an idea can be come up with.

I don't currently have any better ideas on how to handle it than my previous thoughts, but I can keep grabbing the seasons manually for now and if I think of something I think could work I will bring it to you. Hopefully there's some other people that this bothers that might have an idea haha. I wish this one was as simple as get better indexers.

Cheers

**Comment by SF4L** [comment]
Didn't see this issue before coming to Discord today.

My 2 cents: either per-show, per-indexer or advanced global setting approaches would make more sense than comparisons to original release date. I've run into issues where seasons are available on pre-release on a streaming service _before_ they air on the "official" network according to TVDB (obviously, this is a super rare case).

This happens somewhat frequently for BBC (UK) and Stan (AU) series. I've tried talking to TVDB staff about it, but they insist that the original airing network is the network broadcast, and then the streaming service like BBC iPlayer/Stan is a "pre-release" and not to alter the Airdate on TVDB.

**Comment by obsidianwings** [comment]
> Didn't see this issue before coming to Discord today.
> 
> My 2 cents: either per-show, per-indexer or advanced global setting approaches would make more sense than comparisons to original release date. I've run into issues where seasons are available on pre-release on a streaming service _before_ they air on the "official" network according to TVDB (obviously, this is a super rare case).
> 
> This happens somewhat frequently for BBC (UK) and Stan (AU) series. I've tried talking to TVDB staff about it, but they insist that the original airing network is the network broadcast, and then the streaming service like BBC iPlayer/Stan is a "pre-release" and not to alter the Airdate on TVDB.

Yeah, I have quite a bit of commonwealth content so not particularly uncommon for me. It's correct according to the rules so they are definitely not going to change anything. Unfortunately I have not been able to come up with a better idea to implement it than I already suggested, so have just been sticking to doing it manually. 

I'd still be interested in this becoming a reality though.

**Comment by obsidianwings** [comment]
> perhaps allowing season packs if they dropped within X days of the first episode would be a solid middle ground?
> 
> Agree that removing the check is going to cause more problems than not, so a more nuanced feature would be warranted

I just realised I totally misread this the first time around. It looks like I thought you meant if all episodes "aired" with a certain timeframe of the first episode, based on my response. But you're clearly talking about the season packs dropping within the timeframe.

They tend to be available same day as the first episode drops, so this would 100% fix the issue, except for the rare case I recently had where the season pack dropped months before the "official" release, but that's a much rarer problem to deal with manually.

**Comment by vargabp** [comment]
Same issue plagues Futurama, Season 10.
All episodes were released for streaming 15th of September, but not all aired on TV yet.

Wasted plenty of time before I found this was the issue.
TVDB (the only source supported in Sonarr currently) shows the air date for the episodes, hence Sonarr thinks the episodes cannot possibly exist, despite being available on the indexer(s).
As TMDB's date shows the release date as the 15th of September for all of the episodes, if [TMDB support was considered](https://forums.sonarr.tv/t/tmdb-as-metadata-source/32403), it would probably work as another solution to this issue.

**Comment by CraftyClown** [comment]
This is a huge problem in the UK with the BBC releasing everything as full seasons on iPlayer at the same time as they release the first episode on live TV. This means that I only receive the first episode of any new show, while the season pack gets rejected and subsequent episodes aren't uploaded to the trackers as the season pack already exists. It has greatly reduced Sonarr's functionality for me as I have to regularly look up the new shows and manually add the packs... which I inevitably forget to do :(

**Comment by timonsku** [comment]
I had this issue where the metadata was unavailable for the airdate so Sonarr is unable to ever grab the season. Having any option at all to fix this even if its global for now would be great.
Though putting it into a release profile seems like a perfectly reasonable solution. Would a PR be accepted for that?

**Comment by chodeus** [comment]
Just opened sonarr and realised I have been affected by this problem too. Episode 1 of G'Wed s03  was grabbed nearly 3 weeks ago when it aired. The season pack didn't grab because of this issue, and no more episodes have been uploaded as the season pack exists. In the 2 years this issue has been open are we any closer to finding a suitable place to put a toggle to opt-in/out of this behaviour?

I'm with Markus who mentioned delay profiles could be a good place for this. In my mind it makes sense to put it there.

**Comment by obsidianwings** [comment]
Just wanting to check an approach before putting together a PR.

@markus101 mentioned Release Profiles a while back as a possible fit, but the sticking point was what happens when a series has more than one profile applying to it. Turns out the AirDateRestriction/AirDateGracePeriod stuff that got added to Release Profiles recently already dealt with that, so I just copied that pattern:

- New checkbox on Release Profiles: AllowSeasonPackWithoutAllEpisodesAired.
- Resolved the same way AirDateRestriction is — by tags + indexer.
- If a series has multiple profiles that disagree, the strictest one wins (so this can only loosen the check if every applicable profile agrees to loosen it).
- Off by default, and there's a warning in the UI that you should only turn this on for indexers you actually trust, since it's basically saying "believe this release even if it looks incomplete."

Is this a good way to do it or should I look at something different?

**Comment by markus101** [comment]
@obsidianwings I think your suggested approach makes sense, at the moment I can't think of anything additional we'd need for that.
